modules/dnn/src/layers/einsum_layer.cpp

```diff
@@ -202,12 +188,198 @@ bool IsTransposeRequired(size_t input_rank, const std::vector<size_t>& permutati
+    }
+
+    // Compute total number of higher-dimensional slices
+    int total_slices = input.size[0];
+
+    // Create the output matrix (keeping the original dimensions but changing the size of the last dimension to 1)
+    std::vector<int> output_dims(input.dims);
+    for (int i = 0; i < input.dims - 1; ++i) {
+        output_dims[i] = input.size[i];
+    }
+    output_dims.back() = 1;
+    Mat output = Mat(output_dims, input.type());
+
+    int inner_stride = input.size[input.dims - 1];
+    for (int slice = 0; slice < total_slices; ++slice) {
+        for (int j = 0; j < inner_stride; ++j) {
+            // Extract diagonal elements
+            output.at<T>(slice, j, 0) = input.at<T>(slice, j, j);
+        }
+    }
+
+    // Reshape output back to original shape
+    original_dims[rank - 1] = 1;  // Set the last dimension to 1, as we have extracted the diagonal
+    output = output.reshape(1, original_dims);
+    return output;
+}
+
+/* Extract the diagonal elements from the last two dimensions of the tensor.
+For instance, given an input_shape of [1, 2, 3, 3]:
+
+The flexibility in this implementation allows one to choose which of the two
+last dimensions retains its value, determined by the `preserve_innermost_dim_val` parameter.
+
+When preserve_innermost_dim_val == true:
+    The resulting shape is [1, 2, 1, 3], indicating the diagonal has 3 elements,
+    and it keeps the dimension value of the innermost dimension.
+
+When preserve_innermost_dim_val == false:
+    The resulting shape is [1, 2, 3, 1], indicating the diagonal also has 3 elements,
+    but it retains the dimension value of the penultimate dimension. */
+Mat DiagonalInnermostDims(const Mat& input, bool preserve_innermost_dim_val) {
+    const MatShape input_dims = shape(input);
+    int rank = input_dims.size();
+
+    // This is an internal method and we already have finished all validations in the calling method.
+    // We proceed without duplicating all validations again here.
+
+    // We have a minimalistic check here to make sure the innermost dims have the same dim value
+    // as the calling method may have done a transpose before calling this method
+    CV_CheckEQ(input.size[rank - 1], input.size[rank - 2],
+        "innermost dims should have the same dim value to parse the diagonal elements");
+
+    MatShape output_dims;
+    output_dims.reserve(rank);
+
+    for (size_t i = 0; i < rank - 2; ++i) {
+        output_dims.push_back(input_dims[i]);
+    }
+
+    if (preserve_innermost_dim_val) {
+        output_dims.push_back(1);
+        output_dims.push_back(input_dims[rank - 1]);
+    } else {
+        output_dims.push_back(input_dims[rank - 1]);
+        output_dims.push_back(1);
+    }
+
+    // TODO: hande different types
+    Mat output = DiagonalDataAssignment<float>(input);
+
+    if (output_dims != shape(output)){
+        CV_Error(Error::StsError, "Output shape does not match with calculated shape");
+    }
```

Comment:
Do we really need `output_dims` as it's used only for the check?

Reply:
do we not need to check?

Reply:
We probably initialize output somewhere once in the code with such shapes.

Reply:
output shape for this particular function is not saved anywhere. However, overall output shape is save, of course. Because of this I need to check the outputs for compatibility.